auth/bitbucket/server/provider.go

```diff
@@ -0,0 +1,66 @@
+package server
+
+import (
+	"github.com/concourse/atc/auth/verifier"
+	"github.com/dghubble/oauth1"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	"net/http"
+	"time"
+)
+
+type Provider struct {
+	*oauth1.Config
+	verifier.Verifier
+	secrets map[string]string
+}
+
+func (p *Provider) AuthCodeURL(state string, opts ...oauth2.AuthCodeOption) string {
+	requestToken, requestSecret, err := p.Config.RequestToken()
+	if err != nil {
+		panic(err)
+	}
+	authorizationURL, err := p.Config.AuthorizationURL(requestToken)
+	if err != nil {
+		panic(err)
+	}
+	p.secrets[requestToken] = requestSecret
+	return authorizationURL.String()
+}
+
+func (p *Provider) Client(ctx context.Context, t *oauth2.Token) *http.Client {
+	return p.Config.Client(ctx, &oauth1.Token{Token: t.AccessToken, TokenSecret: t.RefreshToken})
+}
+
+func (p *Provider) Exchange(ctx context.Context, code string) (*oauth2.Token, error) {
+	r := ctx.Value("request").(*http.Request)
```

Comment:
Sneaking this through the context seems...sneaky. But I think I see why it was done.

How much work would it involve to just change the interface to take the request as an argument? Other providers would then just take the `code` form value?

Is it too much of a stretch to force v1 and v2 down the same interface?

I wouldn't consider these questions blockers, but I'd like to call it out.

Reply:
So the other providers would take `code` from the `Request` themselves? That probably shouldn't be too hard. I'll give it a try.

We tried to avoid changing the interface, because we don't know how many people rely on it. A similar thing happened with the v1 vs. v2 routes.

Personally, I'd like to either change the v2 route so something like `/oauth/v2/` or pass information about the version through the Elm code, so we can decide which protocol to follow in the callback handler (and get rid of the second set of routes in the process).

Reply:
Yeah - let's switch it to `*http.Request`. All the things that rely on it should be within the ATC, so the scope is limited.

I think there's definitely something off about mixing v1 and v2 like this, but I think we should maybe pull this in and tackle that ourselves (unless you want to give it a go), with #202 in mind as well. With both of these merged, we'd end up with 4 auth flows:

* basic auth
* LDAP basic auth
* oAuth v2
* oAuth v1

We've generalized oAuth v2 with a common interface, but it doesn't seem like that was enough. We may need to make things even more dynamic so that even basic auth is no longer hardcoded.

Reply:
I'll try to have a look at #202 and come up with something, but I won't be too sad if this gets pulled in after changing the interface, either.